fedimint-core/src/db/mod.rs

```diff
@@ -385,13 +392,17 @@ impl Database {
     ///
     /// This function panics when the given number of maximum attempts is zero.
     /// `max_attempts` must be greater or equal to one.
-    pub async fn autocommit<'s: 'dt, 'dt, F, T, E>(
+    pub async fn autocommit<'s, 'dbtx, F, T, E>(
         &'s self,
         tx_fn: F,
         max_attempts: Option<usize>,
     ) -> Result<T, AutocommitError<E>>
     where
-        for<'a> F: Fn(&'a mut DatabaseTransaction<'dt>) -> BoxFuture<'a, Result<T, E>>,
+        's: 'dbtx,
+        for<'r, 'o> F: Fn(
+            &'r mut DatabaseTransaction<'o>,
+            PhantomBound<'dbtx, 'o>,
```

Comment:
I can not get rid of it, but I'm not certain it's necessary.

Reply:
I think the reason why `DatabaseTransactionRef` worked before is that the `into_non_commitable` is introducing an extra lifetime which introduces the problem described in the article linked in `PhantomBound` doc comment, but it's a bit hard to reason about.

Reply:
After we land I might just ask on user forums and maybe someone can confirm / fix.

Reply:
Could we have a `BorrowMut` impl instead? Since we always seem to pass mut references `DerefMut` would be inconvenient if I'm not mistaken as it requires dereferencing first before creating a reference again?

Reply:
@elsirion I don't understand why. `Deref` and `DerefMut` are the only two operators that are implicit, which makes it tempting to use it for all sorts of automatic conversion. The callee could already take `dbtx: DatabaseTransaction<Cap>`, which would make the call work for both `Committable` and `Uncommittable` and save the caller the conversion from `Commitable` to `Uncommittable`, but that requires extra work and remembering about.


One thing I can immediately improve is introduce `beging_transaction_ro` that returns `DatabaseTransaction<Uncommittable` since RO transactions don't ever need `commit_tx`. That should cut down number of conversion by maybe 75% in the user code, since lots of txes are RO.

Reply:
@elsirion I started introducing `.begin_transaction_ro` and while it works, and does eliminate need for `.into_non_commitable` it's a bit weird that it' still possible to call `insert` etc. on them, even if they technically can't modify anything because the transaction will never commit...

Not sure what to think about it.

Reply:
Maybe it should be `begin_transaction_non_committable` or something like that instead.

Reply:
Also what would be nice is if "committable" wasn't 100 characters long with m and t repeating 5 times. Maybe we could abbreviate everywhere to `nc` (kind of like `ro` for ReadOnly)?

Reply:
I stand corrected, `Deref` magic is stronger than I thought, played around with it a bit and seems like the sensible thing to do indeed. The only annoyance is that you can only deref to exactly one target type, but it seems unlikely that we will want more target types in this case.

It might still be useful to have explicit functions to get committable or non-commitable versions in case we implement generic logic where the type marker cannot be inferred.

Reply:
Actually, does `Deref` even work in this case without some unsafe `transmute`ing? The function sig is `Deref::deref(&self) -> &Self::Target`, so we get a reference and want to change the type of that reference, but cannot create a new struct with the correct type either because that will go out of scope on return. So the only option I see is transmuting the reference? Or am I missing something?

Reply:
> So the only option I see is transmuting the reference? Or am I missing something?

Oh, right. We can't do `Deref` without `transmute`, and I'm too scared to use `transmute`